Applications/Spire/Source/Ui/ResponsiveLabel.cpp

```diff
@@ -0,0 +1,230 @@
+    : m_labels(std::move(labels)),
+      m_text_model(std::make_shared<LocalTextModel>()) {
+  m_text_box = make_label(m_text_model, this);
+  proxy_style(*this, *m_text_box);
+  m_style_connection =
+    connect_style_signal(*m_text_box, [=] { on_text_box_style(); });
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  layout->addWidget(m_text_box);
+  m_list_operation_connection = m_labels->connect_operation_signal(
+    [=] (auto operation) { on_list_operation(operation); });
+  update_current_font();
+  reset_mapped_labels();
+  update_display_text();
+}
+
+const std::shared_ptr<ListModel<QString>>&
+    ResponsiveLabel::get_labels() const {
+  return m_labels;
+}
+
+const std::shared_ptr<TextModel>& ResponsiveLabel::get_current() const {
+  return m_text_box->get_current();
+}
+
+const std::shared_ptr<HighlightModel>& ResponsiveLabel::get_highlight() const {
+  return m_text_box->get_highlight();
+}
+
+QSize ResponsiveLabel::sizeHint() const {
+  return m_text_box->sizeHint();
+}
+
+void ResponsiveLabel::resizeEvent(QResizeEvent* event) {
+  if(is_outside_current_bounds(width())) {
+    update_display_text();
+  }
+  QWidget::resizeEvent(event);
+}
+
+int ResponsiveLabel::get_current_label_pixel_width() const {
+  if(!m_current_mapped_index) {
+    return 0;
+  }
+  return m_mapped_labels.at(*m_current_mapped_index).m_pixel_width;
+}
+
+int ResponsiveLabel::get_next_label_pixel_width() const {
+  if(!m_current_mapped_index) {
+    return 0;
+  } else if(*m_current_mapped_index + 1 >
+      static_cast<int>(m_mapped_labels.size()) - 1) {
+    return QWIDGETSIZE_MAX;
+  }
+  return m_mapped_labels.at(*m_current_mapped_index + 1).m_pixel_width;
+}
+
+int ResponsiveLabel::get_pixel_width(const QString& text) const {
+  return QFontMetrics(m_text_box_font).horizontalAdvance(text);
+}
+
+bool ResponsiveLabel::is_outside_current_bounds(int width) const {
+  return width < get_current_label_pixel_width() ||
+    width >= get_next_label_pixel_width();
+}
+
+void ResponsiveLabel::reset_mapped_labels() {
+  m_current_mapped_index = none;
+  m_mapped_labels.clear();
+  m_mapped_labels.reserve(m_labels->get_size());
+  for(auto i = 0; i < m_labels->get_size(); ++i) {
+    m_mapped_labels.push_back({i, get_pixel_width(m_labels->get(i))});
+  }
+  sort_mapped_labels();
+}
+
+void ResponsiveLabel::set_current(const optional<int> mapped_label_index) {
+  m_current_mapped_index = mapped_label_index;
+  if(!m_current_mapped_index) {
+    m_text_model->set("");
+    return;
+  }
+  m_text_model->set(
+    m_labels->get(m_mapped_labels.at(*m_current_mapped_index).m_list_index));
+  if(m_text_box->is_text_elided() && m_current_mapped_index != 0) {
+    adjustSize();
+    update_display_text();
+  }
+}
+
+void ResponsiveLabel::sort_mapped_labels() {
+  std::sort(m_mapped_labels.begin(), m_mapped_labels.end(),
+    [=] (const auto& first, const auto& second) {
+      return first.m_pixel_width < second.m_pixel_width;
```

Comment:
When sorting the mapped_labels, the index isn't considered. According to the description of  the ResponsiveLabel, the value with the lowest list index among the same length values is displayed.